### PR TITLE
Use Deleter to remove output in a number of more cases

### PR DESCRIPTION
**Pull request by lptr**
This takes https://github.com/gradle/gradle/issues/8684 one step close to completion. This PR removes all the usages of `FileUtils.forceDelete()` and `FileUtils.cleanDirectory()`:

- [x] executing non-incrementally
- [x] when loading from cache
- [x] cleaning up stale outputs
- [x] `SyncCopyAction`
- [x] `AntlrTask`
- [x] `Groovydoc`
- [x] `Javadoc`

### CODE REVIEW

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java [hunk 1]
```diff
@@ -56,31 +56,51 @@ public DefaultDeleter(LongSupplier timeProvider, Predicate<? super File> isSymli
...
-            return true;
+            return deleteRecursively(root, followSymlinks, false);
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean cleanRecursively(File root, boolean followSymlinks) throws IOException {
+        if (root.exists() && shouldRemoveContentsOf(root, followSymlinks)) {
```
Comment (wolfs) [inline subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java]: I think for `cleanRecursively`, we should always delete the contents of the root, even if it is a symlink. If we don't, then the directory won't be empty after this call. We can leave this until we make the general change whether to follow symlinks for deletion or not.

Comment (lptr) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java]: I've introduced `ensureEmptyDirectory()`, PTAL.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java [hunk 2]
```diff
@@ -63,9 +63,17 @@ public boolean deleteRecursively(File root, boolean followSymlinks) throws IOExc
     }
 
     @Override
-    public boolean cleanRecursively(File root, boolean followSymlinks) throws IOException {
-        if (root.exists() && shouldRemoveContentsOf(root, followSymlinks)) {
+    public boolean ensureEmptyDirectory(File root, boolean followSymlinks) throws IOException {
+        if (root.isDirectory()) {
```
Comment (wolfs) [inline subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java]: Don't we need to check if the root is a symlink and remove it? A symlink to a directory still returns `true` for `isDirectory()`.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/Deleter.java [hunk 3]
```diff
@@ -38,9 +38,17 @@
...
-     * Attempts to clean the given directory recursively, removing all of its contents.
+     * Makes sure that the given target is an empty directory.
+     *
+     * If target is...
+     *
+     * <ul>
+     *     <li>a directory, then its contents are removed recursively,</li>
+     *     <li>a symlink pointing to an existing directory, then the linked directory's contents are removed recursively,</li>
+     *     <li>a file, or a symlink to an existing file, it is deleted and a directory is created in its place,</li>
+     *     <li>non-existent, then a directory is created in its place.</li>
```
Comment (wolfs) [inline subprojects/files/src/main/java/org/gradle/internal/file/impl/Deleter.java]: I don't think the code in `DefaultDeleter` creates a directory if the file is non-existant.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java (hunk 2 above)
Comment (lptr) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java]: We don't if we adhere to the Javadoc, which says that we keep root symlinks pointing to directories.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/Deleter.java (hunk 3 above)
Comment (lptr) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/Deleter.java]: You're faster than CI! :) Fixed.

File: subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java (hunk 2 above)
Comment (wolfs) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java]: Ah, I got confused by the two Javadocs talking about symlinks. For the directory we have
```
a symlink pointing to an existing directory, then the linked directory's contents are removed recursively,
```

IIUC, currently the linked directory's contents are only removed if `followSymlinks` is `true`, right? Is this intended?

Comment (lptr) [reply subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java]: Good catch. Fixed it and added a test.

**Pull request merged**